exporter/collector/metrics.go

```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
+// continually retry the same request until success. Other non-retryable errors
+// drop metrics.
+func (me *MetricsExporter) readWALAndExport(ctx context.Context) error {
+	for {
+		select {
+		case <-me.shutdownC:
+			return nil
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		bytes, err := me.wal.Read(me.rWALIndex.Load())
+		if err == nil {
+			req := new(monitoringpb.CreateTimeSeriesRequest)
+			if err = proto.Unmarshal(bytes, req); err != nil {
+				return err
 			}
 
-			// always record the number of successful points
-			recordPointCountDataPoint(ctx, succeededPoints, "OK")
-			if failedPoints > 0 {
-				recordPointCountDataPoint(ctx, failedPoints, st)
+			err = me.export(ctx, req)
+			if err != nil {
+				me.obs.log.Warn(fmt.Sprintf("error exporting to GCM: %+v", err))
+			}
+			// retry at same read index if retryable (network) error
+			s := status.Convert(err)
+			if s.Code() == codes.DeadlineExceeded || s.Code() == codes.Unavailable {
+				me.obs.log.Error("non-retryable error, skipping request")
+				continue
 			}
+
+			err = me.wal.TruncateFront(me.rWALIndex.Load())
 			if err != nil {
-				errs = append(errs, fmt.Errorf("failed to export time series to GCM: %+v", s))
+				return err
 			}
+			// move read index forward if non retryable error (or exported successfully)
+			me.rWALIndex.Add(1)
+			continue
+		}
+
+		// ErrNotFound from wal.Read() means the index is either 0 or out of
+		// bounds (indicating we're probably at the end of the WAL). That error
+		// will trigger a file watch for new writes (below this). For other
+		// errors, fail.
+		if !errors.Is(err, wal.ErrNotFound) {
+			return err
+		}
+
+		// Must have been ErrNotFound, start a file watch and block waiting for updates.
+		if err := me.watchWAL(ctx); err != nil {
+			return err
 		}
 	}
-	if len(errs) > 0 {
-		return multierr.Combine(errs...)
+}
+
+// watchWAL watches the WAL directory for a write then returns to the
+// continuallyPopWAL() loop.
+func (me *MetricsExporter) watchWAL(ctx context.Context) error {
+	walWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	err = walWatcher.Add(me.walPath)
+	if err != nil {
+		return err
+	}
+
+	watchCh := make(chan error)
+	var wErr error
+	go func() {
+		defer func() {
+			watchCh <- wErr
+			close(watchCh)
+			walWatcher.Close()
+		}()
+
+		select {
+		case <-me.shutdownC:
+			return
+		case <-ctx.Done():
+			wErr = ctx.Err()
+			return
+
+		case event, ok := <-walWatcher.Events:
+			if !ok {
+				return
+			}
+			switch event.Op {
+			case fsnotify.Remove:
+				me.obs.log.Error("WAL file deleted")
+			case fsnotify.Rename:
+				me.obs.log.Error("WAL file renamed")
+			case fsnotify.Write:
+				wErr = nil
+			}
+
+		case eerr, ok := <-walWatcher.Errors:
+			if ok {
+				wErr = eerr
+			}
+		}
+	}()
+	err = <-watchCh
+	return err
+}
+
+func (me *MetricsExporter) walRunner(ctx context.Context) {
+	defer me.goroutines.Done()
+	for {
+		select {
```

Comment:
Do we need to handle `ctx.Done()`?

Reply:
It looks like this context is the one passed to `NewGoogleCloudMetricsExporter()`. Is that one intended for initialization only? We're not even passing it to `go mExp.exportMetricDescriptorRunner()` on L235.

Reply:
`shutdownC` is only closed if `Shutdown()` is called, so I think handling the main context closing is good to have to prevent from blocking on other terminations (non-gracefully?)

Reply:
> so I think handling the main context closing is good to have to prevent from blocking on other terminations (non-gracefully?)

That makes sense, but I wasn't sure of the collector's contract here. I figured it would eventually cancel that "initialization" context which would cause this goroutine to exit but that doesn't seem to be the case.